statsmodels/sandbox/examples/gam_gsoc2015/gam.py

```diff
@@ -0,0 +1,267 @@
+            keep_index = np.nonzero(np.abs(res.params) > 1e-4) [0]
+            rmat = np.eye(len(res.params))[drop_index]
+
+            # calling fit_constrained raise
+            # "RuntimeError: maximum recursion depth exceeded in __instancecheck__"
+            # fit_constrained is calling fit, recursive endless loop
+            if drop_index.any():
+                # todo : trim kwyword doesn't work, why not?
+                #res_aux = self.fit_constrained(rmat, trim=False)
+                res_aux = self._fit_zeros(keep_index, **kwds)
+                return res_aux
+            else:
+                return res
+
+## this class will be later removed and taken from another push
+class Penalty(object):
+    """
+    A class for representing a scalar-value penalty.
+    Parameters
+    wts : array-like
+        A vector of weights that determines the weight of the penalty
+        for each parameter.
+    Notes
+    -----
+    The class has a member called `alpha` that scales the weights.
+    """
+
+    def __init__(self, wts):
+        self.wts = wts
+        self.alpha = 1.
+
+    def func(self, params):
+        """
+        A penalty function on a vector of parameters.
+        Parameters
+x        ----------
+        params : array-like
+            A vector of parameters.
+        Returns
+        -------
+        A scalar penaty value; greater values imply greater
+        penalization.
+        """
+        raise NotImplementedError
+
+    def grad(self, params):
+        """
+        The gradient of a penalty function.
+        Parameters
+        ----------
+        params : array-like
+            A vector of parameters
+        Returns
+        -------
+        The gradient of the penalty with respect to each element in
+        `params`.
+        """
+        raise NotImplementedError
+
+
+class GamPenalty(Penalty):
+    
+    def __init__(self, wts=1, alpha=1, cov_der2=None, der2=None):
+        
+        self.wts = wts #should we keep wts????
+        self.alpha = alpha
+        self.cov_der2 = cov_der2
+        self.der2 = der2
+    
+    def func(self, params):
+        ''' 
+        1) params are the coefficients in the regression model
+        2) der2  is the second derivative of the splines basis
+        '''
+        
+        # The second derivative of the estimated regression function
+        f = np.dot(self.der2, params)
+
+        return self.alpha * np.sum(f**2)
+
+    def grad(self, params):
+        ''' 
+        1) params are the coefficients in the regression model
+        2) der2  is the second derivative of the splines basis
+        3) cov_der2 is obtained as np.dot(der2.T, der2)
+        '''
+                
+        return 2 * self.alpha * np.dot(self.cov_der2, params) 
+    
+    def deriv2(self, params):
+
+        return 2 * self.alpha * self.cov_der2
+
+
+
+class LogitGam(PenalizedMixin, Logit):
+    pass
+
+
+class GLMGam(PenalizedMixin, GLM):
+    pass
+  
+
+
+def test_gam_penalty():
```

Comment:
move it to a test module `test_gam.py` in a `tests` folder so that nose will find it automatically.
nose testing is looking for files/modules and functions or classes that start or end with `test` (and similar)